### PR TITLE
Add support for 'homogenize'

### PR DESCRIPTION
**Pull request by ctjacobs**
Added a function to create a homogeneous version of a DirichletBC. Also comes with a test.

Addresses #443.

### CODE REVIEW

File: tests/regression/test_bcs.py [hunk 1]
```diff
@@ -113,6 +113,16 @@ def test_homogenize_doesnt_overwrite_function(a, u, V, f):
     solve(a == 0, u, bcs=[bc])
     assert abs(u.vector().array()).max() == 0.0
 
+def test_homogenize(V):
+    bc = [DirichletBC(V, 10, 1), DirichletBC(V, 20, 2)]
+
+    homogeneous_bc = bcs.homogenize(bc)
```
Comment (miklos1) [inline tests/regression/test_bcs.py]: Maybe you want to add 'homogenize' to __all__ in bcs.py, so that you don't need to prefix it with the module name.

**Comment by miklos1** [comment]
I history clean up might be great. (I can do that if you have no experience editing history.)
Otherwise, I think it is good to go.

File: firedrake/bcs.py [hunk 2]
```diff
@@ -222,3 +222,17 @@ def zero(self, r):
...
+
+def homogenize(bc):
+    """Create a homogeneous version of a :class:`.DirichletBC` object and return it. If
+    ``bc`` is a list or tuple containing one or more :class:`.DirichletBC` objects,
+    then return a list of the homogeneous versions of those :class:`.DirichletBC`s.
+
+    :arg bc: a :class:`.DirichletBC`, or list or tuple comprising :class:`.DirichletBC`(s).
+    """
+    if isinstance(bc, (list, tuple)):
+        return [homogenize(i) for i in bc]
```
Comment (wence-) [inline firedrake/bcs.py]: I would write this as:

```python
try:
    return [homogenize(i) for i in bc]
except TypeError:
    # not iterable
    return DirichletBC(bc.function_space(), 0, bc.sub_domain)
```

That way you treat any iterable.

**Comment by ctjacobs** [comment]
Thanks. Lawrence's suggestion has been applied and all commits have been squashed into one.

**Comment by ctjacobs** [comment]
Travis is green. Is this good to go?

**Pull request merged**